tests/unit/format/DBase4Test.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+/*
+ * To change this license header, choose License Headers in Project Properties.
+ * To change this template file, choose Tools | Templates
+ * and open the template in the editor.
+ */
+
+namespace org\majkel\dbase\format;
+
+use org\majkel\dbase\Format;
+use org\majkel\dbase\tests\utils\AbstractFormatTest;
+use org\majkel\dbase\Field;
+
+/**
+ * Record class tests
+ *
+ * @author majkel
+ *
+ * @coversDefaultClass \org\majkel\dbase\format\DBase3
+ */
+class DBase4Test extends AbstractFormatTest {
+
+    const CLS = '\org\majkel\dbase\format\DBase4';
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getFormatObject() {
+        return $this->mock(self::CLS)->new();
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getSupportedTypes() {
+        return array(
+            Field::TYPE_CHARACTER, Field::TYPE_DATE, Field::TYPE_LOGICAL,
+            Field::TYPE_MEMO, Field::TYPE_NUMERIC, Field::TYPE_FLOAT
+        );
+    }
+
+    /**
+     * @covers ::createHeader
+     */
+    public function testCreateHeader() {
+        $format = $this->getFormatObject();
+        $header = $this->reflect($format)->createHeader($this->getHeaderData(array(
+            'v' => 4,
+        )));
+        self::assertTrue($header->isValid());
+    }
+
+    /**
+     * @covers ::createHeader
+     */
+    public function testCreateHeaderUnknownFormat() {
+        $format = $this->getFormatObject();
+        $header = $this->reflect($format)->createHeader($this->getHeaderData(array(
+            'v' => 666,
+        )));
+        self::assertTrue($header->isValid());
```

Comment:
Please investigate why this test fails. The only case where the header could be invalid is when the calculated records count does not match file size

Reply:
I see what is going on here.
It is using bitwise logic here.
return $header->setValid($header->getVersion() & $this->getVersion());
So with dBase4 I need to use 668 for the test since it will be doing 668 & 4 in the above setValid method. And 4 in binary is 100. I have to choose a number that has its bit flipped in the 3rd spot and 666 doesn't do this but 668 does.
668 is ‭0010 1001 1100‬ in binary.